lib/schemaToQueryTree.js

```diff
@@ -120,39 +119,87 @@ function getSkipKey(fieldTypeDefinition, field, parentFieldTypeDefinition) {
   return `${field.name}-${fieldTypeDefinition.name}-${parentFieldTypeName}`;
 }
 
-module.exports = function buildQueryTreeFromField(field, typeDictionary, skipped = [], parentFieldTypeDefinition = null) {
-  const fieldTypeName = magiclyExtractFieldTypeName(field);
-  const fieldTypeDefinition = typeDictionary[fieldTypeName];
+const getQueryFieldsModes = {
+  QUERYABLE_FIELDS: 'QUERYABLE_FIELDS',
+  ALL_FIELDS: 'ALL_FIELDS'
+};
 
-  // this are base invocations for all subsequent queries
-  const queriesForRootField = getFieldNameOrExamplesIfNecessary(field);
-  if (fieldTypeDefinition.fields === null) { // isLeafType
-    if(queriesForRootField.length === 0) {
-      return null;
+module.exports = {
+  getQueryFieldsModes: getQueryFieldsModes,
+  getQueryFields: function getQueryFields(mode, field, typeDictionary, visitedFields, isRoot = true, parentFieldTypeDefinition = null) {
+    let coverageMagic = [];
+    const fieldTypeName = magiclyExtractFieldTypeName(field);
+    const fieldTypeDefinition = typeDictionary[fieldTypeName];
+    if (isRoot) {
+      coverageMagic.push(`${fieldTypeName}___${field.name}`);
+      if (mode === getQueryFieldsModes.QUERYABLE_FIELDS) {
+        const queriesBecauseRootAndWeDontKnowIfWeShouldStart = getFieldNameOrExamplesIfNecessary(field);
+        if (queriesBecauseRootAndWeDontKnowIfWeShouldStart.length === 0) {
+          return [];
+        }
+      }
     }
 
-    return queriesForRootField.join(' ');
-  }
-
-  if (fieldTypeDefinition.kind === 'OBJECT') {
-    skipped.push(getSkipKey(fieldTypeDefinition, field, parentFieldTypeDefinition));
-  }
+    if (fieldTypeDefinition.fields === null) { // isLeafType
+      return coverageMagic;
+    }
 
-  let queryNode = null;
-  const allFields = getFields(field, typeDictionary);
+    if (fieldTypeDefinition.kind === 'OBJECT') {
+      visitedFields.push(getSkipKey(fieldTypeDefinition, field, parentFieldTypeDefinition));
+    }
 
-  _.forIn(allFields, (childField) => {
-    const childFieldTypeName = magiclyExtractFieldTypeName(childField);
-    const childFieldType = typeDictionary[childFieldTypeName];
+    const allFields = getFields(field, typeDictionary);
+    _.forIn(allFields, (childField) => {
+      const childFieldTypeName = magiclyExtractFieldTypeName(childField);
+      const childFieldType = typeDictionary[childFieldTypeName];
+      const noOfPossibleQueries = getFieldNameOrExamplesIfNecessary(childField).length;
+
+      if (
+        (noOfPossibleQueries > 0 || mode === getQueryFieldsModes.ALL_FIELDS) &&
+        visitedFields.indexOf(getSkipKey(childFieldType, childField, fieldTypeDefinition)) === -1
```

Comment:
Maybe a helper function (`hasChildFieldBeenVisited(childFieldType, childField, fieldTypeDefiniton)`) ?

I regret writing that.